BUILD.bazel

```diff
@@ -75,6 +76,7 @@ go_context_data(
         "//go/platform:internal_cgo_off": None,
         "//conditions:default": ":cgo_context_data",
     }),
+    test_init = "//go/tools/test_init",
```

Comment:
This should only be an implicit dependency of `go_test`.

`coverdata` is a little different, because every `go_library` might need to import it. `test_init` only needs to be imported by `go_test` archives.

Reply:
Sorry for being so dense, but how do I do that exactly?  I tried doing that first, but I could never figure out exactly how and where to add this dep without following the coverdata pattern.  I can try hacking at this more and give more detail on exactly what I ran into.

Reply:
`_testmain_additional_srcs` is a better pattern to follow. Add an attribute to `go_test` that starts with `_`. It should be an `attr.label` with a default value pointing to that `go_library` target.

Reply:
I think I'm close, but I'm missing something.  I added this:

And this:

```diff
--- a/go/private/rules/test.bzl
+++ b/go/private/rules/test.bzl
@@ -137,12 +137,14 @@ def _go_test_impl(ctx):
         resolve = None,
     )
     test_deps = external_archive.direct + [external_archive]
-    test_deps.append(go.test_init)
+    print("test_deps[0]:", type(test_deps[0]), test_deps[0])
+    print("_testmain_additional_deps[0]:", type(ctx.files._testmain_additional_deps[0]), ctx.files._testmain_additional_deps[0])
+    print("go.coverdata:", type(go.coverdata), go.coverdata)
     if ctx.configuration.coverage_enabled:
         test_deps.append(go.coverdata)
     test_source = go.library_to_source(go, struct(
         srcs = [struct(files = [main_go] + ctx.files._testmain_additional_srcs)],
-        deps = test_deps,
+        deps = test_deps + ctx.files._testmain_additional_deps,
     ), test_library, False)
     test_archive, executable, runfiles = go.binary(
         go,
@@ -201,6 +203,10 @@ _go_test_kwargs = {
             default = ["@io_bazel_rules_go//go/tools/testwrapper:srcs"],
             allow_files = go_exts,
         ),
+        "_testmain_additional_deps": attr.label_list(
+            providers = [GoLibrary],
+            default = ["@io_bazel_rules_go//go/tools/testinit"],
+        ),
         # Workaround for bazelbuild/bazel#6293. See comment in lcov_merger.sh.
         "_lcov_merger": attr.label(
             executable = True,
```

But the thing is, when I get to the print statements, the _testmain_additional_deps[0] is a File, whereas go.coverdata is a struct.  I can't figure out how to make it be the kind of thing it needs to be. :/

Reply:
Resolved all the feedback except for this-- I'm a little out of my depth. :(

Reply:
Alright.. I finally stumbled my way through this!  I needed `ctx.attr._testmain_additional_deps` to get the actual target, not `ctx.files._testmain_additional_deps` which only has the output file.  Big oof!